File: services/zk-authz/bridge/bridge.go
// Package bridge implements the deterministic SHA-256 -> BN254 field bridge
// defined in docs/GNARK_BINDING.md. It MUST agree bit-for-bit with the Python
// side (cvztte.zkauthz.bridge); substitution tests enforce this.
package bridge

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/big"
	"strings"
)

// CapDomain is the domain-separation tag for capability field encoding.
const CapDomain = "cvztte.cap.v3"

// DigestToLimbs splits a 32-byte digest into two big-endian 128-bit limbs.
func DigestToLimbs(d []byte) (*big.Int, *big.Int, error) {
	if len(d) != 32 {
		return nil, nil, fmt.Errorf("digest must be 32 bytes, got %d", len(d))
	}
	hi := new(big.Int).SetBytes(d[:16])
	lo := new(big.Int).SetBytes(d[16:])
	return hi, lo, nil
}

// Sha256HexToLimbs parses "sha256:<hex>" (or bare 64-hex) into (hi, lo).
func Sha256HexToLimbs(commit string) (*big.Int, *big.Int, error) {
	h := strings.TrimPrefix(commit, "sha256:")
	raw, err := hex.DecodeString(h)
	if err != nil {
		return nil, nil, fmt.Errorf("invalid sha256 hex: %w", err)
	}
	return DigestToLimbs(raw)
}

// CapabilityField encodes a capability name to a single (<2^128) field integer:
// int_be( sha256( CapDomain || 0x00 || name )[16:32] ).
func CapabilityField(name string) *big.Int {
	h := sha256.New()
	h.Write([]byte(CapDomain))
	h.Write([]byte{0x00})
	h.Write([]byte(name))
	sum := h.Sum(nil)
	return new(big.Int).SetBytes(sum[16:])
}

File: cvztte/zkauthz/bridge.py
"""Deterministic SHA-256 -> BN254 field bridge (docs/GNARK_BINDING.md).

MUST agree bit-for-bit with the Go ``bridge`` package. Substitution tests
(tests/integration/test_zkauthz.py) assert this on shared vectors by comparing
against the ``zk-authz encode`` subcommand.
"""

from __future__ import annotations

import hashlib

from cvztte.errors import CVZTTEError, ErrorCode

# Domain-separation tag for capability field encoding (must equal Go CapDomain).
CAP_DOMAIN = "cvztte.cap.v3"


def digest_to_limbs(digest: bytes) -> tuple[int, int]:
    """Split a 32-byte digest into two big-endian 128-bit limbs (hi, lo)."""
    if len(digest) != 32:
        raise CVZTTEError(
            ErrorCode.SCHEMA_INVALID,
            "digest must be 32 bytes",
            detail={"length": len(digest)},
        )
    hi = int.from_bytes(digest[:16], "big")
    lo = int.from_bytes(digest[16:], "big")
    return hi, lo


def sha256_to_field_pair(commit: str) -> tuple[int, int]:
    """Parse ``"sha256:<hex>"`` (or bare 64-hex) into an ordered (hi, lo) pair."""
    raw = commit[len("sha256:") :] if commit.startswith("sha256:") else commit
    try:
        digest = bytes.fromhex(raw)
    except ValueError as exc:
        raise CVZTTEError(
            ErrorCode.SCHEMA_INVALID, "invalid sha256 hex", detail={"value": commit}
        ) from exc
    return digest_to_limbs(digest)


def capability_field(name: str) -> int:
    """Encode a capability name to a single (<2^128) field integer:

    ``int_be( sha256( CAP_DOMAIN || 0x00 || name )[16:32] )``.
    """
    h = hashlib.sha256()
    h.update(CAP_DOMAIN.encode("utf-8"))
    h.update(b"\x00")
    h.update(name.encode("utf-8"))
    return int.from_bytes(h.digest()[16:], "big")
